src/adb.js

```diff
@@ -1 +1,393 @@
+	 *
+	 * @param {String} deviceId - The id of the device or emulator.
+	 * @param {String} apkFile - The application apk file to install.
+	 * @param {Object} [opts] - Install options.
+	 * @returns {Promise}
+	 * @access public
+	 */
+	installApp(deviceId, apkFile, opts = {}) {
+		if (typeof deviceId !== 'string' || !deviceId) {
+			return Promise.reject(new TypeError('Expected device ID to be a string.'));
+		}
+
+		apkFile = util.expandPath(apkFile);
+		if (!util.existsSync(apkFile)) {
+			return Promise.reject(new TypeError(`APK file "${apkFile}" does not exist.`));
+		}
+
+		return this
+			.devices()
+			.then(devices => {
+				if (!devices.some(d => d.id === deviceId)) {
+					return Promise.reject(new Error('Device not found.'));
+				}
+			})
+			.then(sdk.detect)
+			.then(result => util.run(result.sdk.executables.adb, ['-s', deviceId, 'install', '-r', apkFile]));
+	}
+
+	/**
+	 * Starts an application on the specified device/emulator.
+	 *
+	 * @param {String} deviceId - The id of the device or emulator.
+	 * @param {String} appId - The application's id.
+	 * @param {String} activity - The name of the activity to run.
+	 * @returns {Promise}
+	 * @access public
+	 */
+	startApp(deviceId, appId, activity) {
+		if (typeof deviceId !== 'string' || !deviceId) {
+			return Promise.reject(new TypeError('Expected device ID to be a string.'));
+		}
+
+		if (typeof appId !== 'string' || !appId) {
+			return Promise.reject(new TypeError('Expected app ID to be a string.'));
+		}
+
+		if (typeof activity !== 'string' || !activity) {
+			return Promise.reject(new TypeError('Expected activity name to be a string.'));
+		}
+
+		return this.shell(deviceId, `am start -n ${appId} /. ${activity.replace(/^\./, '')}`);
+	}
+
+	/**
+	 * Stops an application on the specified device/emulator.
+	 *
+	 * @param {String} deviceId - The id of the device or emulator.
+	 * @param {String} appId - The application's id.
+	 * @returns {Promise}
+	 * @access public
+	 */
+	stopApp(deviceId, appId) {
+		if (typeof deviceId !== 'string' || !deviceId) {
+			return Promise.reject(new TypeError('Expected device ID to be a string.'));
+		}
+
+		if (typeof appId !== 'string' || !appId) {
+			return Promise.reject(new TypeError('Expected app ID to be a string.'));
+		}
+
+		return this
+			.getPid(deviceId, appId)
+			.then(pid => this.shell(deviceId, `am force-stop ${appId}`));
+	}
+
+	/**
+	 * Retrieves a list of connected devices and emulators.
+	 *
+	 * @returns {Promise}
+	 * @access public
+	 */
+	devices() {
+		return new Connection(this)
+			.exec('host:devices')
+			.then(data => this.parseDevices(data));
+	}
+
+	/**
+	 * Streams output from logcat into the specified handler until the adb logcat
+	 * process ends.
+	 *
+	 * @param {String} deviceId - The id of the device or emulator.
+	 * @returns {EventEmitter} Emits events `message`, `error` and `close`.
+	 * @access public
+	 */
+	logcat(deviceId) {
+		const emitter = new EventEmitter;
+		sdk
```

Comment:
For readability sake, do an explicit `Promise.resolve()` to kick off the promise chain:

```
Promise.resolve()
	.then(() => sdk.detect(this.opts))
	.then(result => {
		//
	});
```